### ISSUE TITLE
Is there a way to prevent a value from being converted into a SqlParameter?

### ISSUE THREAD

**Issue by matthewjones555** [linq2db/linq2db#1350]
I'm trying to work around an issue in SqlCe 3.0, where you cannot use a SQL parameter with a db function, such as COALESCE, or ISNULL.

Is there a way of telling LinqToDb to not parameterize a specific value?

I've been digging into the source, but was wondering if there was a built in way of doing this, before I look at making a fork.

### Steps to reproduce
Use a query that includes a coalesce on a SQL CE database.

```c#
var query = from c in context.MyTable
            where (c.StartDate ?? SqlDateTime.MinValue.Value) <= DateTime.Now
            select c;
```

This produces the following SQL

```SQL
-- MyContext SqlCe
DECLARE @p1 DateTime
SET     @p1 = '1753-01-01'
DECLARE @p2 DateTime
SET     @p2 = '2018-10-17 10:07:02.353'

SELECT
	[t1].[Id],
	[t1].[FirstName],
	[t1].[LastName],
	[t1].[StartDate]
FROM
	[MyTable] [t1]
WHERE
	Coalesce([t1].[StartDate], @p1) <= @p2
```

Here's the exception that running the query produces.

```
Exception message: System.Data.SqlServerCe.SqlCeException: 'The specified argument value for the function is not valid. [ Argument # = 2,Name of function(if known) = coalesce ]'
Stack trace:
   at System.Data.SqlServerCe.SqlCeCommand.CompileQueryPlan()
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteCommand(CommandBehavior behavior, String method, ResultSetOptions options)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior)
   at LinqToDB.Data.DataConnection.ExecuteReader()
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteReader()
   at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__9`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

If I manually modify the SQL to the following, it executes successfully.  This is what I'd like to achieve with LinqToSql if possible.

```SQL
-- MyContext SqlCe
DECLARE @p1 DateTime
SET     @p1 = '2018-10-17 10:07:02.353'

SELECT
	[t1].[Id],
	[t1].[FirstName],
	[t1].[LastName],
	[t1].[StartDate]
FROM
	[MyTable] [t1]
WHERE
	Coalesce([t1].[StartDate], '1753-01-01') <= @p1
```

### Environment details
linq2db version: MASTER, commit 9bcf33e23d9be452c39bfc47203e3b165450aad7
Database Server: SQL CE 3.0
Database Provider: SQL CE 3.0
Operating system: Windows 10
Framework version: NET Framework 4.5.1

**Comment by sdanyliv** [comment]
@matthewjones555, i'm out of computer to check this in detail, but it is possible. 
1. You can inline parameters in whole query: `context.InlineParameters = true;`
2. You can write new function in some class and use in queries:
```cs
[Sql.Function(InlineParameters = true)]
public static void T Coalesce<T>(T? value, T defaultValue) where T: struct;
{
   throw new NotImplementedException();
}
```
3. Wait for fix in new version, or create PR. Looking at realization, and see that this can be done easily in `SqlCeSqlOptimizer`

**Comment by matthewjones555** [comment]
Excellent!  Thank you for the speedy reply.  I will look at this and see if I can figure it out.

**Comment by matthewjones555** [comment]
FYI, I've made the change, as detailed in option 3 from @sdanyliv's post.  The pull request is here: https://github.com/linq2db/linq2db/pull/1352

I'm happy to take on whatever feedback you guys have, and make any changes you'd like to see.

I've tried to make the change in a logical place, but I'm largely unfamiliar with the inner workings of LinqToDB.

Thanks in advance for any feedback.

Matt

**Comment by sdanyliv** [comment]
@matthewjones555, thanks for the PR. Looks good. But i need to check how it works with cache (try to call the same expression several times with different parameters) and write some tests (it is manadatory). 
Making PR should be like a game and i suppose digging into linq2db sources should give more understanding what library can do.

**Comment by igor-tkachev** [comment]
The ToSql method should be able to inline particular parameters.

**Comment by matthewjones555** [comment]
Thanks for the feedback guys.

In my current use case, I'd rather not use Linq2Db specific helpers, like Sql.ToSql, as I have some generic queries that can work for any provider, so I can't use provider specific syntax, because I have to maintain compatibility.  

It'd be nice if the behaviour could be driven from a setting, or better yet from the current SQL provider. 
 Do you still want me to work on adding a setting for the PR?  Or shall I just keep this as my own private fork?

**Comment by sdanyliv** [comment]
Setting will be appreciated. We will handle this our-self. Thanks again.

**Comment by MaceWindu** [comment]
Fixed by #1352
